Pip: use 'pip list' when available for package list

rcutmore opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

Pip
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.2.0 (add-pip-test c28741035d) last updated 2016/09/01 08:12:12 (GMT -400)
  lib/ansible/modules/core: (issue-3060 30c0e9dcc2) last updated 2016/09/01 07:47:15 (GMT -400)
  lib/ansible/modules/extras: (devel e8a5442345) last updated 2016/09/01 07:59:52 (GMT -400)
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

This updates the pip module to use `pip list` when available to get a list of installed packages. If `pip list` is not available then `pip freeze` is used instead. This allows certain packages, such as pip itself, to be checked, which normally aren't listed in the results of `pip freeze`.
Fixes #3060

<!---
If you are fixing an existing issue, please include "Fixes #nnnn" in your commit
message and your description; but you should still explain what the change does.
-->

<!--- Paste verbatim command output below, e.g. before and after your change -->

Command:

```
- name: check for pip package in check_mode
  pip: name=pip virtualenv=testenv state=present
  check_mode: True
```

Before:

```
changed: [testhost] => {"changed": true, "cmd": "/home/rcutmore/ansible_testing/pipenv/bin/pip freeze", "stderr": "", "stdout": "ansible==2.2.0\n-e git+https://github.com/dvarrazzo/pyiso8601@3237e6267f7c73858ad5b62d27824070f068b252#egg=iso8601\n", "stdout_lines": ["ansible==2.2.0", "-e git+https://github.com/dvarrazzo/pyiso8601@3237e6267f7c73858ad5b62d27824070f068b252#egg=iso8601"]}
```

After:

```
ok: [testhost] => {"changed": false, "cmd": "/home/rcutmore/ansible_testing/pipenv/bin/pip list", "stderr": "", "stdout": "ansible (2.2.0)\niso8601 (0.1.5, /home/rcutmore/ansible_testing/pipenv/src/pyiso8601)\npip (8.1.2)\nsetuptools (26.1.1)\nwheel (0.29.0)\n", "stdout_lines": ["ansible (2.2.0)", "iso8601 (0.1.5, /home/rcutmore/ansible_testing/pipenv/src/pyiso8601)", "pip (8.1.2)", "setuptools (26.1.1)", "wheel (0.29.0)"]}
```

rcutmore committed:
Pip: handle parsing different pip commands

rcutmore committed:
Pip: use 'pip list' when available

rcutmore commented:
Associated test for these changes added [here](https://github.com/ansible/ansible/pull/17360).

gregdek commented:
Thanks @rcutmore. To the current maintainers, @mattupstate please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

abadger commented on packaging/language/pip.py:
```diff
@@ -413,15 +433,14 @@ def main():
         if requirements or has_vcs:
-            freeze_cmd = '%s freeze' % pip
-            out_freeze_before = module.run_command(freeze_cmd, cwd=chdir)[1]
+            _, _, out_freeze_before, _ = _get_packages(module, pip, chdir)
```
Need to add some error checking here.  Before, wasn't robust but it probably did the job.  It checked that the 3-tuple of output was the same before and after.  Now, we're only saving stdout so we're not capturing whether there was an error condition or the package just has no installable packages to report.  Need to be explicit about checking the rc value now (and probably reporting stdout+stderr in the fail_json msg)

abadger commented on packaging/language/pip.py:
```diff
@@ -443,7 +462,7 @@ def main():
                     changed = 'Successfully installed' in out_pip
                 else:
-                    out_freeze_after = module.run_command(freeze_cmd, cwd=chdir)[1]
+                    _, _, out_freeze_after, _ = _get_packages(module, pip, chdir)
```
Error checking here as well.

abadger commented on packaging/language/pip.py:
```diff
@@ -204,13 +204,36 @@ def _get_full_name(name, version=None):
     for pkg in installed_pkgs:
-        if '==' not in pkg:
+        # Package listing will be different depending on which pip
+        # command was used ('pip list' vs. 'pip freeze').
```
Since `_get_packages()` returns the command it used (pip freeze vs pip list) can't we pass that information in here so that we can separately parse freeze output and list output?  That seems less error prone as pip's output changes in the future and other people have to update the conditionals in this function.

abadger commented:
Thanks @rcutmore.  Made a few comments but overall this looks like a good strategy to fix the issue.  If I don't respond after you've updated the code feel free to ping me on IRC.  I'm abadger1999 on irc.freenode.net #ansible-devel channel.

rcutmore committed:
Pip: explicitly check which command is used

rcutmore committed:
Pip: add error checking when fetching packages

rcutmore commented:
@abadger thanks for reviewing. I pushed a few changes, let me know whether or not this addresses the issues raised or if I'm missing something.
- I updated `_is_present()` to make the parsing of `pip list` and `pip freeze` expicit.
- I updated `_get_packages()` to fail if there's an error with both `pip list` and `pip freeze`. From what I can tell `pip freeze` will return a result code of 0 even when there are no packages installed.

It looks like the previous code was just checking stdout before and after (`out_freeze_before = module.run_command(freeze_cmd, cwd=chdir)[1]` and `out_freeze_after = module.run_command(freeze_cmd, cwd=chdir)[1]`). I can update this to check the 3 tuple output if you think that would be preferable.

gregdek commented:
@rcutmore A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

rcutmore commented:
ready_for_review

gregdek commented:
Thanks @rcutmore. To the current maintainers, @mattupstate please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

abadger commented:
Thank you @rcutmore !  Merged to devel for inclusion in 2.2.

abadger commented:
@rcutmore Just letting you know, when I merged the test cases I realized that the code still had the old behaviour if pip was too old to have pip list.  It took me a few tries but I finally was able to get a slower fallback method working.  Commits to implement the fallback are here: https://github.com/ansible/ansible-modules-core/commit/5b1994cb42ad52a3a3cbf5461cbe6f8af2a69262  and here: https://github.com/ansible/ansible-modules-core/commit/b4f1d0d0c2f8e580e972c75860a9e44dba89e3d8

if you want to take a look and make any corrections if needed.

rcutmore commented:
@abadger Those commits make sense and look good to me. When I have a bit more time I'll take a closer look to be sure. Thanks again for reviewing (and cleaning up) my changes!
